extensions/amp-ad-network-base/0.1/amp-ad-network-base.js

```diff
@@ -0,0 +1,265 @@
+import {sendXhrRequest} from './amp-ad-utils';
+
+const TAG = 'amp-ad-network-base';
+
+/**
+ * @abstract
+ */
+export class AmpAdNetworkBase extends AMP.BaseElement {
+
+  constructor(element) {
+    super(element);
+
+    /** @private {Object<string, !./amp-ad-type-defs.Renderer>} */
+    this.renderers_ = map();
+
+    /** @private {Object<string, !./amp-ad-type-defs.Validator>} */
+    this.validators_ = map();
+
+    /** @private {Object<string, string>} */
+    this.recoveryModes_ = map();
+
+    /** @private {?./amp-ad-type-defs.LayoutInfoDef} */
+    this.initialSize_ = null;
+
+    /** @const @private {!AmpAdContext} */
+    this.context_ = new AmpAdContext(this.win);
+
+    /**
+     * When true, indicates that the renderer and validator should not be
+     * freed (e.g., for refreshable implementations).
+     * @private {boolean}
+     */
+    this.isReusable_ = false;
+
+    // Register default error modes.
+    for (const failureType in FailureType) {
+      this.recoveryModes_[failureType] = RecoveryModeType.COLLAPSE;
+    }
+
+    /**
+     * Number of times to retry a failed request. Zero by default.
+     * @type {number}
+     */
+    this.retryLimit_ = 0;
+  }
+
+  /**
+   * @param {string} failure
+   * @param {string} recovery
+   * @final
+   */
+  onFailure(failure, recovery) {
+    if (this.recoveryModes_[failure]) {
+      dev().warn(TAG,
+          `Recovery mode for failure type ${failure} already registered!`);
+    }
+    this.recoveryModes_[failure] = recovery;
+  }
+
+  /**
+   * @param {!./amp-ad-type-defs.Validator} validator
+   * @param {string=} type
+   * @final
+   */
+  registerValidator(validator, type = 'default') {
+    if (this.validators_[type]) {
+      dev().warn(TAG, `${type} validator already registered.`);
+    }
+    this.validators_[type] = validator;
+  }
+
+  /**
+   * @param {!./amp-ad-type-defs.Renderer} renderer
+   * @param {string} type
+   * @final
+   */
+  registerRenderer(renderer, type) {
+    if (this.renderers_[type]) {
+      dev().warn(TAG, `Rendering mode already registered for type '${type}'`);
+    }
+    this.renderers_[type] = renderer;
+  }
+
+  /**
+   * @return {string} The finalized ad request URL.
+   * @protected
+   * @abstract
+   */
+  getRequestUrl() {
+    // Subclass must override.
+  }
+
+  /** @param {boolean} isReusable */
+  setIsReusable(isReusable) {
+    this.isReusable_ = isReusable;
+  }
+
+  /** @param {number} retries */
+  setRequestRetries(retries) {
+    this.retryLimit_ = retries;
+  }
+
+
+  /**
+   * Processes the ad response as soon as the XHR request returns.
+   * @param {?../../../src/service/xhr-impl.FetchResponse} response
+   * @private
+   */
+  handleAdResponse_(response) {
+    if (!response.arrayBuffer) {
+      this.handleFailure_(FailureType.INVALID_RESPONSE);
+      return;
+    }
+    response.arrayBuffer().then(unvalidatedBytes => {
+      const validatorType = response.headers.get('validator-type') || 'default';
```

Comment:
We need a better name for the response header:
tentatively: 'validator-type' -> 'AMP-Ad-Response-Type'

Reply:
Done.